### ISSUE TITLE
What about Bayes By Backpropagation?

### ISSUE THREAD

**Issue by ferrine** [Lasagne/Lasagne#699]
Hello, 
What about adding a mixin for making layer bayesian?

The following idea is described [here](http://arxiv.org/pdf/1505.05424v2.pdf)
Some code example of it is provided [here](https://github.com/DTU-deeplearning/day4-VAE/blob/master/bayes%20by%20backprop.ipynb) by @DTU-deeplearning or [here](https://gist.github.com/rocknrollnerd/c5af642cf217971d93f499e8f70fcb72) by @rocknrollnerd

**Comment by benanne** [comment]
PRs are welcome, if anyone is interested in giving this a go :) I think it might be a challenge to write something that is general enough to apply to any layer, though.

**Comment by rocknrollnerd** [comment]
Hi! I've been trying to wrap my code around some layer-like abstractions, actually, though I'm not sure if BBB is well-recognized enough for a general-purpose library like Lasagne. There's [parmesan](https://github.com/casperkaae/parmesan) maintained by people from DTU and specifically dedicated to variational neural network methods; maybe it's better to PR them?

**Comment by Ferrine** [comment]
Yes, they might have been interested in it. When I'll have a bit more time, I'll write them too. 
BTW, could you please describe the experience of your attempt?

**Comment by casperkaae** [comment]
Hi @Ferrine 

I'm one of the maintainers of Parmesan. We would like a PR on BBB there if it turns out to be to difficult to fit into lasagne :) Maybe it would also fit well with the lasagne recipes?

We made an implementation for our summerschool last summer you might find usefull (https://github.com/DTU-deeplearning/day4-VAE/blob/master/bayes%20by%20backprop_solution.ipynb). As @rocknrollnerd mentioned it is not completely trivial to implement in a clean general way and can also be a bit tricky to get to converge (pay attention to the exact initializations ) 

Cheers
Casper

**Comment by rocknrollnerd** [comment]
@Ferrine 

Nothing fancy, just some OOP styling. :-) I'll post it in a [pastebin](http://pastebin.com/raw/ECgnW9jw) but it's far from complete to submit a PR, I'm afraid. Actually I was trying a rather bizarre combination of BBB and Generative adversarial networks framework at the moment, so I just needed some classes that could act like stacking layers. I've also made a Tensorflow implementation basically repeating line-by-line the one you've posted, in case anyone's interested.

Speaking of bad design, I guess first place goes to my implementation of a sampling loop: it should definitely use `theano.scan`, not the simple `for` loop. As @casperkaae have mentioned, designing an objective is tricky (all the minibatch reweighting stuff). I'm afraid at the moment I don't feel confident enough to try to make a decent contribution, either to Lasagne or Parmesan, but I will be happy to assist you, if you choose to do it instead. :-) (you can contact me by public email in my profile)

And also nice to see you, @casperkaae, and thanks for the great summer school time. :-) I've actually been among the students and it was Charles' lecture on BBB that have got me interested.

**Comment by Ferrine** [comment]
@rocknrollnerd

I think that metaclass will be the easiest way to implement this feature. Some methods will be decorated a bit and I hope it will be very flexible. As for objective, It is a puzzle for me now, but it is real to make it in general way I think. I'll create a branch in my forked Lasange with some not tested code.

I'll need some help as it's my first contribution experience. Will be happy to work in team:)

**Comment by Ferrine** [comment]
@benanne 

I have some idea based on introspection when computing graph. I saw that `deterministic` graph is realised in get_output_for method of layer. That gave me one hypothesis, but it will fail if one condition is not satisfied. It is following: all trainable shared variables appear in the graph only when get_output_for is called. Does this condition hold? If is, I see possibility to change them before future computations.

**Comment by Ferrine** [comment]
@f0k

Hmm, yes, that's good news for me. Thanks. 
What about BatchNormLayer, that will be no problem for me.

**Comment by Ferrine** [comment]
Issue [#693](https://github.com/Lasagne/Lasagne/issues/693)  is closely related with mine. I think they can be merged.

**Comment by f0k** [comment]
> I think they can be merged.

github cannot merge issues, and it's not the same -- that issue was only about using Lasagne to construct a model for PyMC3.

**Comment by Ferrine** [comment]
Is it a problem if my implementation will not support deterministic output?

**Comment by f0k** [comment]
For a Lasagne Recipe it's not a problem at all -- just make sure to document it so people don't get trapped by this.

**Comment by Ferrine** [comment]
@rocknrollnerd

I have some strange results. Network without qw-pw converges fine, but when I add it, there is no hope to converge.
Here is 15k iterations without variational cost:

**Std of predicted probability**
![img](https://pp.vk.me/c615722/v615722630/1bf93/5OHo-Bpek6o.jpg)

But when I add qw - pw it yelds something ugly
![img](https://pp.vk.me/c615722/v615722630/1c392/0nVBx83VBzE.jpg)

With bad loss path
![image](https://pp.vk.me/c615722/v615722630/1c399/EXnN-VtzeI8.jpg)

More code is in the [gist](https://gist.github.com/Ferrine/d195d4e40ea454148a35ed9a9f2e8b35)
and [notebook](https://github.com/Ferrine/Lasagne/blob/dists/examples/bbp.ipynb)

**Comment by Ferrine** [comment]
Finally got it!
## Usage example:
### Before
```python
l_in = lasagne.layers.InputLayer(X.shape)
l1_hidden = DenseLayer(l_in, num_units=N_HIDDEN, W=Normal(1), b=Normal(1),
                       nonlinearity=lasagne.nonlinearities.tanh)
# and so on
# ...
# ok, we have net
# let's get objective!
true_output = T.ivector('true_output')
net_output = lasagne.layers.get_output(l_output).ravel()
all_params = lasagne.layers.get_all_params(l_output)
objective = lasagne.objectives.binary_crossentropy(net_output, true_output)
objective = objective.sum()
# train it as you wish
```
### What I propose
```python
from lasagne.layers.bayes import bbpwrap, NormalApproximation, Accumulator

# 1) choose your favorite Layer
# 2) wrap your favorite Layer
@bbpwrap(NormalApproximation(pstd=1))
class BayesDenseLayer(DenseLayer):
    pass
# Gracias! It's bayesian!

# 3) create the thing that will do some dirty work for us
#    It will collect all variational cost
acc = Accumulator()

l_in = lasagne.layers.InputLayer(X.shape)
# 4) pass `acc` as first argument
l1_hidden = BayesDenseLayer(acc, l_in, num_units=N_HIDDEN, W=Normal(1), b=Normal(1),
                            nonlinearity=lasagne.nonlinearities.tanh)
# also possible
# myW = {'mu':Normal(1), 'rho':Normal(1)}
# l1_hidden = BayesDenseLayer(acc, l_in, num_units=N_HIDDEN, W=myW, b=Normal(1),
#                             nonlinearity=lasagne.nonlinearities.tanh)
# and so on
# ...
# ok, we have our net
# as usual we define objective
true_output = T.ivector('true_output')
net_output = lasagne.layers.get_output(l_output).ravel()
all_params = lasagne.layers.get_all_params(l_output)
objective = lasagne.objectives.binary_crossentropy(net_output, true_output)
objective = objective.sum()

# 5) that thing has collected the variational cost
#   all we need is to get it and add to our objective
varcost = acc.get_cost()
objective += varcost
# It's better to train with adam
updates = lasagne.updates.adam(objective, all_params)
```
So, it appears that decorating Layers is very easy. We have for instance bbp conv/lstm layers or VAE out of the box
BTW you cannot get deterministic output but you can sample about 10 or whatever you want samples of predictions and calculate theirs mean and std
```python
sampled_proba = theano.function([l_in.input_var], net_output)
def predict_proba(x, steps=100):
    mean = 0
    for _ in range(steps):
        mean += sampled_proba(x)
    return mean / steps

def predict_std(x, steps=100):
    cum = list()
    for _ in range(steps):
        cum.append(sampled_proba(x))
    return np.std(cum, 0)
```
Calculating them can be very useful. You can answer some real world questions like "what's tha chance that prediction is wrong?". Or you can calculate mean "confidence" for dataset and decide if it is overconfident or not.
### Results
![bbp](https://cloud.githubusercontent.com/assets/11705326/17383966/f912cba2-59e9-11e6-81d6-995fa8d6924b.png)
### Training process
![train_cost](https://cloud.githubusercontent.com/assets/11705326/17384044/4ae5f67a-59ea-11e6-88ec-5143a050213b.png)
![test_err](https://cloud.githubusercontent.com/assets/11705326/17384047/4ea558f0-59ea-11e6-80af-125735654312.png)
### PS
Now It's nearly ready for pull request. Please decide if it suits Lasagne or not. I still can have it in my repo as a tiny wrapping.

Source code is available [here](https://gist.github.com/Ferrine/a003ace716c278ab87669f2fbd37727b).

**Comment by f0k** [comment]
> Please decide if it suits Lasagne or not. I still can have it in my repo as a tiny wrapping.

Just to document it here as well, in #731 we agreed that it would be a better fit for [Parmesan](/casperkaae/parmesan).

**Comment by ferrine** [comment]
BTW, I found that all I wanted to do can be easily implemented with [PyMC3](https://github.com/pymc-devs/pymc3) and there is no need in any wrapper even! See this great [tutorial](http://twiecki.github.io/blog/2016/07/05/bayesian-deep-learning/) by @twiecki

**Comment by twiecki** [comment]
@ferrine Glad that was helpful, I was surprised by how easy it was. Would be curious to see/hear more about your use-case.

**Comment by ferrine** [comment]
@twiecki That idea came to after reading this [paper](http://arxiv.org/pdf/1505.05424.pdf), I was really impressed. As I could not find any reasonable implementation I decided to do it by myself. I have no use-case yet because I'm still studying. I hope it will be useful in my future diploma in economics.

**Comment by twiecki** [comment]
Interesting @ferrine. Were you able to add the variational free energy?

**Comment by ferrine** [comment]
@twiecki Yes, I've done some [cheat decorations](https://gist.github.com/Ferrine/a003ace716c278ab87669f2fbd37727b) over Layer class and kept tracking all attempts to create a param. It worked nice and I reproduced the [first part](http://twiecki.github.io/blog/2016/06/01/bayesian-deep-learning/) of your tutorial using my wrapper. I did not publish that notebook, but you can see the API and results above. I trained the network with ADAM and it seems to be converging faster than Adagrad in PyMC3. 
Another difference was about reparametrization trick.
I used
`W=mean + log(1+exp(rho))*e`
instead of
`W=mean + exp(rho)*e`
It is more numerically stable

**Comment by f0k** [comment]
> `log(1+exp(rho))`

That's the [softplus function](http://deeplearning.net/software/theano/library/tensor/nnet/nnet.html#theano.tensor.nnet.nnet.softplus). Interesting observation. I found `exp()` to learn better for constraining parameters to positive values, but in a very different setting.

**Comment by ferrine** [comment]
I can check the difference for variational tasks soon using toy example above
